notebooks/examples/python/Ship2ShipTransfers/02. Overlap Detection.py

```diff
@@ -0,0 +1,162 @@
+# Databricks notebook source
+# MAGIC %md # Overlap Detection
+# MAGIC
+# MAGIC We now try to detect potentially overlapping pings using a buffer on a particular day.
+
+# COMMAND ----------
+
+from pyspark.sql.functions import *
+import mosaic as mos
+
+spark.conf.set("spark.databricks.labs.mosaic.geometry.api", "ESRI")
+spark.conf.set("spark.databricks.labs.mosaic.index.system", "H3")
+mos.enable_mosaic(spark, dbutils)
+
+
+# COMMAND ----------
+
+cargos_indexed = spark.read.table("ship2ship.cargos_indexed").filter(
+    col("BaseDateTime").between(
+        "2018-01-31T00:00:00.000+0000", "2018-01-31T23:59:00.000+0000"
+    )
+)
+display(cargos_indexed)
+cargos_indexed.count()
+
+# COMMAND ----------
+
+# MAGIC %md ## Buffering
+# MAGIC
+# MAGIC 1. Convert the point into a polygon by buffering it with a certain area to turn this into a circle.
+# MAGIC 2. Index the polygon to leverage more performant querying.
+# MAGIC
+# MAGIC ![](http://1fykyq3mdn5r21tpna3wkdyi-wpengine.netdna-ssl.com/wp-content/uploads/2018/06/image14-1.png)
+# MAGIC
+# MAGIC
+# MAGIC
+# MAGIC Since our projection is not in metres, we convert from decimal degrees, with `(0.00001 - 0.000001)` as being equal to one metre at the equator.
+# MAGIC Here we choose an buffer of roughly 100 metres.
+# MAGIC Ref: http://wiki.gis.com/wiki/index.php/Decimal_degrees
+
+# COMMAND ----------
+
+one_metre = 0.00001 - 0.000001
+buffer = 100 * one_metre
+
+(
+    cargos_indexed.repartition(sc.defaultParallelism * 20)
+    .withColumn("buffer_geom", mos.st_buffer("point_geom", lit(buffer)))
+    .withColumn("ix", mos.mosaic_explode("buffer_geom", lit(9)))
+    .write.mode("overwrite")
+    .saveAsTable("ship2ship.cargos_buffered")
+)
+
+# COMMAND ----------
+
+# DBTITLE 1,We can optimise our table to colocate data and make querying faster
+# MAGIC %sql
+# MAGIC OPTIMIZE ship2ship.cargos_buffered ZORDER BY (ix.index_id, BaseDateTime);
+# MAGIC SELECT * FROM ship2ship.cargos_buffered;
+
+# COMMAND ----------
+
+# MAGIC %md ## Implement Algorithm
+
+# COMMAND ----------
+
+buffered_events = spark.read.table("ship2ship.cargos_buffered").repartition(
+    sc.defaultParallelism * 20
```

Comment:
Now `cargos_buffered` is a delta table, so I would not expect a repartition to be necessary. If there is a reason for it, could you please add it in a comment?

Reply:
Without a repartition, it all runs it on a single the executor. The default execution plan is just bad. 
![Screenshot 2022-07-26 at 08 36 11](https://user-images.githubusercontent.com/17589078/181007314-b86f5208-6b38-4481-99eb-2ed197d4c36f.png)

I will add a comment to specify!

Reply:
Oh, interesting!
The data is too small perhaps. I'm happy with the comment. :)